### PR TITLE
IWAID-6 BE - CRUD Citas Médicas

### PR DESCRIPTION
**Pull request by DavidVelaz17**
### Ticket:
* [IWAID-6](https://iwaids.atlassian.net/browse/IWAID-6?atlOrigin=eyJpIjoiYjAzYjcwZjhhMjIwNDY5NDk3ODZjZWJiOWEwMzNiZDUiLCJwIjoiaiJ9)

### What Changed:
* Changed server port from 8080 to 8081
* New RESTful endpoint for handling medical appointments

### How To Test:
*  To test each endpoint you need to excecute the project and then go to Postman.
    * To get All appointments, create a new GET request with the following route: http://localhost:8081/iwaid/appointments/
    * To get a specific appointment, create a new GET request with the following route: http://localhost:8081/iwaid/appointments/{id}
    * To create an appointment then create a POST request with the following route:http://localhost:8081/iwaid/appointments/ 
       * Go to "Body " section and select "raw", then in "text" select "JSON" and pase the following:
          * {
    "patientId":"1",
    "doctorId":"2",
    "date":"2023-11-02",
    "hour":"FIVE_PM",
    "notes":"Patient need to arrive seven minutes earlier"
}
    * To delete an appointment, create a DELETE request with the following route: http://localhost:8081/iwaid/appointments/{id}
    * To update an appointment, crete a PATCH request with the following route: http://localhost:8081/iwaid/appointments/{id}
        * Go to "Body " section and select "raw", then in "text" select "JSON" and pase the following:
             *  {
    "patientId":"1",
    "doctorId":"1",
    "date":"2024-11-02",
    "hour":"TRHEE_PM",
    "notes":"Patient need to arrive five minutes earlier"
}
        > If the route has an {id} then you must change it for a number, ex: http://localhost:8081/iwaid/appointments/2

### Evicence: 
* Get All Appointments:
findAll
* Fin Appointment By Id:
FindById
* Create an Appointment:
create
* Delete Appointment By Id:
delete
* Update Appointment By Id:
update

### CODE REVIEW

File: consultorio-medico/target/classes/com/iwa/iwaid/consultoriomedico/dto/AppointmentDTO$AppointmentDTOBuilder.class
Comment (SendyCruz) [inline consultorio-medico/target/classes/com/iwa/iwaid/consultoriomedico/dto/AppointmentDTO$AppointmentDTOBuilder.class:1]: no cargar estos archivos

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/restcontrollers/AppointmentController.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    @GetMapping("/{appointmentId}")
+    public ResponseEntity<AppointmentDTO> getAppointmentById(@PathVariable("appointmentId") final int appointmentId) throws Exception {
+        final AppointmentDTO appointmentDTO = appointmentService.getAppointmentById(appointmentId);
+        return ResponseEntity.ok().body(appointmentDTO);
+    }
+
+    @PostMapping("/")
+    public ResponseEntity saveAppointment(@RequestBody @Valid AppointmentForm form) throws Exception {
```
Comment (SendyCruz) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/restcontrollers/AppointmentController.java]:
```suggestion
    public ResponseEntity createAppointment(@RequestBody @Valid AppointmentForm form) throws Exception {
```

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/AppointmentService.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+    public AppointmentDTO getAppointmentById(final int id) throws Exception {
+        validateIfAppointmentExists(id);
+        final Appointment appointment = appointmentRepository.findById(id).get();
+        final DoctorDTO doctorDTO = doctorService.getDoctorById(appointment.getDoctorId());
+        final PatientDTO patientDTO = patientService.getPatientById(appointment.getPatientId());
+        return AppointmentDTO.build(appointment, doctorDTO, patientDTO);
+    }
+
+    public AppointmentDTO saveAppointment(final AppointmentForm form) throws Exception {
```
Comment (SendyCruz) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/AppointmentService.java]:
```suggestion
    public AppointmentDTO createAppointment(final AppointmentForm form) throws Exception {
```

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/convertors/HourToIntConvertor.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package com.iwa.iwaid.consultoriomedico.convertors;
+
+import com.iwa.iwaid.consultoriomedico.entity.Hour;
+import jakarta.persistence.AttributeConverter;
+import jakarta.persistence.Converter;
+
+@Converter(autoApply = true)
+public class HourToIntConvertor implements AttributeConverter<Hour,Integer> {
```
Comment (cgarcia-iwa) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/convertors/HourToIntConvertor.java]:
```suggestion
public class HourConvertor implements AttributeConverter<Hour,Integer> {
```

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/form/AppointmentForm.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    @NotNull
+    private int doctorId;
+
+    @ApiObjectField(name = "date", description = "Appointment's date")
+    @NotNull
+    @DateTimeFormat(pattern = "yyyy-MM-dd")
+    @Future
+    private LocalDate date;
+
+    @ApiObjectField(name = "hour", description = "Appointment's hour")
```
Comment (SendyCruz) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/form/AppointmentForm.java:30]: @NotNull

**Review by SendyCruz — CHANGES_REQUESTED** [review]

**Comment by SendyCruz** [comment]
Duda: como se va a controlar, que ya no se pueda agendar una cita en la misma fecha y hora?

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/entity/Appointment.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+    @Column(name = "doctors_Id", nullable = false)
+    private int doctorId;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "doctors_Id", nullable = false, insertable = false, updatable = false)
+    private Doctor doctor;
+
+    @Column(name = "date")
+    private LocalDate date;
```
Comment (cgarcia-iwa) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/entity/Appointment.java:46]: que vas a guardar, dependiendo la respuesta verificar si es correcto el tipo LocalDate o Date

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/restcontrollers/AppointmentController.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+import java.util.List;
+
+@RestController
+@RequestMapping(path = "/iwaid/appointments")
+@RequiredArgsConstructor
+public class AppointmentController {
+
+    private final AppointmentService appointmentService;
+
+    @GetMapping("/")
```
Comment (cgarcia-iwa) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/restcontrollers/AppointmentController.java]:
```suggestion
    @GetMapping
```

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/restcontrollers/AppointmentController.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    public List<AppointmentDTO> getAllAppointments() {
+        final List<AppointmentDTO> appointments = appointmentService.getAll();
+        return appointments;
```
Comment (cgarcia-iwa) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/restcontrollers/AppointmentController.java]:
```suggestion
        return appointmentService.getAll();
```

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/restcontrollers/AppointmentController.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    @GetMapping("/{appointmentId}")
+    public ResponseEntity<AppointmentDTO> getAppointmentById(@PathVariable("appointmentId") final int appointmentId) throws Exception {
+        final AppointmentDTO appointmentDTO = appointmentService.getAppointmentById(appointmentId);
+        return ResponseEntity.ok().body(appointmentDTO);
+    }
+
+    @PostMapping("/")
```
Comment (cgarcia-iwa) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/restcontrollers/AppointmentController.java]:
```suggestion
    @PostMapping
```

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/restcontrollers/AppointmentController.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    public ResponseEntity saveAppointment(@RequestBody @Valid AppointmentForm form) throws Exception {
+        return ResponseEntity.status(HttpStatus.CREATED).body(appointmentService.saveAppointment(form));
+    }
+
+    @DeleteMapping("/{appointmentId}")
+    public ResponseEntity<AppointmentDTO> deleteAppointment(@PathVariable("appointmentId") final int appointmentId) throws Exception {
```
Comment (cgarcia-iwa) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/restcontrollers/AppointmentController.java]:
```suggestion
    public ResponseEntity<Void> deleteAppointment(@PathVariable("appointmentId") final int appointmentId) throws Exception {
```

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/AppointmentService.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+    private final DoctorService doctorService;
+    private final PatientService patientService;
+    private final AppointmentRepository appointmentRepository;
+
+    public List<AppointmentDTO> getAll() {
+        final List<Appointment> appointments = appointmentRepository.findAll();
+        final Map<Integer, DoctorDTO> doctorDTOMap = getDoctorsMap(appointments.stream().map(Appointment::getDoctorId).toList());
+        final Map<Integer, PatientDTO> patientDTOMap = getPatientsMap(appointments.stream().map(Appointment::getPatientId).toList());
+        return appointments.stream().map(appointment -> AppointmentDTO.build(appointment, doctorDTOMap.get(appointment.getDoctorId()), patientDTOMap.get(appointment.getPatientId()))).toList();
```
Comment (cgarcia-iwa) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/AppointmentService.java]: dale un formato de cascada

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/AppointmentService.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+    private void validateIfAppointmentExists(int id) throws Exception {
+        if (!appointmentRepository.existsById(id)) {
+            throw new Exception();
+        }
+    }
+
+    private Map<Integer, DoctorDTO> getDoctorsMap(List<Integer> doctorsIds) {
+        final Map<Integer, DoctorDTO> doctorDTOMap = doctorService.getDoctorsByIds(doctorsIds);
+        return doctorDTOMap;
```
Comment (cgarcia-iwa) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/AppointmentService.java]:
```suggestion
        return doctorService.getDoctorsByIds(doctorsIds);
```

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/DoctorService.java [hunk 12]
```diff
@@ -46,7 +49,13 @@ public DoctorDTO updateDoctor(final DoctorForm form, final int id) throws Except
         return DoctorDTO.build(doctor);
     }
 
-    private void validateIfDoctorExists(int id) throws Exception {
+    public Map<Integer, DoctorDTO> getDoctorsByIds(final List<Integer> doctorIds) {
+        final List<Doctor> doctors = doctorRepository.findAllById(doctorIds);
+        final List<DoctorDTO> doctorDTOs = doctors.stream().map(DoctorDTO::build).toList();
```
Comment (cgarcia-iwa) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/DoctorService.java]: mueve la linea 54 y 53 a un metodo privado que funcione igual que el findAll y lo mandas a traer aqui

**Review by cgarcia-iwa — CHANGES_REQUESTED** [review]
Subiste los archivos de target, quitalos y agrega la carpeta a la exepcion

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/dto/AppointmentDTO.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+@Data
+@Builder
+public class AppointmentDTO {
+    @ApiObjectField(name = "id", description = "Appointment's id")
+    private int id;
+
+    @ApiObjectField(name = "patientId", description = "Patient's id")
+    private int patientId;
+
+    @ApiObjectField(name = "patient", description = "Patient's data")
```
Comment (cgarcia-iwa) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/dto/AppointmentDTO.java]: el nombre de la variable ya no coincide con el api object

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/dto/AppointmentDTO.java [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+    private PatientDTO patientDTO;
+
+    @ApiObjectField(name = "doctorId", description = "Doctor's id")
+    private int doctorId;
+
+    @ApiObjectField(name = "doctor", description = "Doctor's data")
```
Comment (cgarcia-iwa) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/dto/AppointmentDTO.java]: ditto

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/restcontrollers/AppointmentController.java [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+    @DeleteMapping("/{appointmentId}")
+    public ResponseEntity<Void> deleteAppointment(@PathVariable("appointmentId") final int appointmentId)
+            throws Exception {
+        appointmentService.deleteAppointment(appointmentId);
+        return ResponseEntity.ok().build();
+    }
+
+    @PatchMapping("/{appointmentId}")
+    public ResponseEntity<AppointmentDTO> updateAppointmentById(@RequestBody @Valid AppointmentForm form,
+                                                                @PathVariable("appointmentId") final int appointmentId)
```
Comment (cgarcia-iwa) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/restcontrollers/AppointmentController.java]: cooregir a formato cascada mas adecuado

**Review by cgarcia-iwa — CHANGES_REQUESTED** [review]

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/DoctorService.java [hunk 16]
```diff
@@ -46,9 +49,18 @@ public DoctorDTO updateDoctor(final DoctorForm form, final int id) throws Except
...
 
-    private void validateIfDoctorExists(int id) throws Exception {
+    public Map<Integer, DoctorDTO> getDoctorsByIds(final List<Integer> doctorIds) {
+        final List<Doctor> doctors = doctorRepository.findAllById(doctorIds);
+        return doctorDTOs(doctors);
+    }
+
+    public void validateIfDoctorExists(final int id) throws Exception {
         if (!doctorRepository.existsById(id)) {
             throw new Exception("Doctor not found");
```
Comment (cgarcia-iwa) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/DoctorService.java]: Todos los errores guardarlo en un archivos de constantes, para evitar que el usar el mismo error en otras secciones pueda variar

**Review by cgarcia-iwa — CHANGES_REQUESTED** [review]

**Review by cgarcia-iwa — APPROVED** [review]

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/AppointmentService.java [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+                .map(appointment -> AppointmentDTO
+                     .build(appointment,
+                            doctorDTOMap
+                            .get(appointment.getDoctorId()),
+                            patientDTOMap
+                            .get(appointment.getPatientId())))
+                .toList();
+    }
+
+    public AppointmentDTO getAppointmentById(final int id) throws Exception {
```
Comment (SendyCruz) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/AppointmentService.java]:
```suggestion
    public AppointmentDTO getAppointmentById(final int appointmentId) throws Exception {
```

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/AppointmentService.java [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+                appointment.getPatientId(),
+                appointment.getDate(),
+                appointment.getHour());
+        doctorService.validateIfDoctorExists(appointment.getDoctorId());
+        patientService.validateIfPatientExist(appointment.getPatientId());
+        appointmentRepository.save(appointment);
+        return AppointmentDTO.build(appointment);
+    }
+
+    public void deleteAppointment(final int id) throws Exception {
```
Comment (SendyCruz) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/AppointmentService.java]:
```suggestion
    public void deleteAppointment(final int appointmentId) throws Exception {
```

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/AppointmentService.java [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+                appointment.getDate(),
+                appointment.getHour());
+        doctorService.validateIfDoctorExists(appointment.getDoctorId());
+        patientService.validateIfPatientExist(appointment.getPatientId());
+        appointment.updateAppointment(form);
+        appointmentRepository.save(appointment);
+        return AppointmentDTO.build(appointment);
+    }
+
+    private void validateIfAppointmentExists(final int id) throws Exception {
```
Comment (SendyCruz) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/AppointmentService.java]:
```suggestion
    private void validateIfAppointmentExists(final int appointmentId) throws Exception {
```

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/AppointmentService.java [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+        if (!appointmentRepository.existsById(id)) {
+            throw new Exception(messages.getString("not.found") + " -Appointment:" + id);
+        }
+    }
+
+    private void validateDoctorAndDateAndHour(
```
Comment (SendyCruz) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/AppointmentService.java]:
```suggestion
    private void validateDoctorAvailability(
```

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/AppointmentService.java [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+    final LocalDate date,
+    final Hour hour)
+        throws Exception {
+        if (appointmentRepository.existsByDoctorIdAndDateAndHour(doctorId, date, hour)) {
+            throw new Exception(messages.getString("date.taken") + " Doctor:" + doctorId);
+        } else
+            log.info(messages.getString("date.and.hour.available"));
+    }
+
+    private void validatePatientAndDateAndHour(
```
Comment (SendyCruz) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/AppointmentService.java]:
```suggestion
    private void validatePatientAvailability(
```

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/DoctorService.java [hunk 22]
```diff
@@ -46,9 +53,22 @@ public DoctorDTO updateDoctor(final DoctorForm form, final int id) throws Except
         return DoctorDTO.build(doctor);
     }
 
-    private void validateIfDoctorExists(int id) throws Exception {
+    public Map<Integer, DoctorDTO> getDoctorsByIds(final List<Integer> doctorIds) {
+        final List<Doctor> doctors = doctorRepository.findAllById(doctorIds);
+        return doctorDTOs(doctors);
+    }
+
+    public void validateIfDoctorExists(final int id) throws Exception {
```
Comment (SendyCruz) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/DoctorService.java]:
```suggestion
    public void validateIfDoctorExists(final int doctorId) throws Exception {
```

File: consultorio-medico/src/main/resources/ValidationMessages.properties [hunk 23]
```diff
@@ -3,4 +3,7 @@ not.null.field=This field cannot be left empty
 name.right.length=The minimum length for this field is 3 characters and the maximum 100
 address.right.length=The minimum length for this field is 3 characters and the maximum 500
 phoneNumber.right.length=The minimum length for this field is 10 characters and the maximum 13
-positive.tel.number=Please insert a real telephone number
\ No newline at end of file
+positive.tel.number=Please insert a real telephone number
```
Comment (SendyCruz) [inline consultorio-medico/src/main/resources/ValidationMessages.properties]:
```suggestion
positive.tel.number= Enter a valid phone number.
```

File: consultorio-medico/src/main/resources/ValidationMessages.properties [hunk 24]
```diff
@@ -3,4 +3,7 @@ not.null.field=This field cannot be left empty
...
+not.found=The requested data does not exist yet
```
Comment (SendyCruz) [inline consultorio-medico/src/main/resources/ValidationMessages.properties]:
```suggestion
not.found=The requested data does not exist yet.
```

File: consultorio-medico/src/main/resources/ValidationMessages.properties [hunk 25]
```diff
@@ -3,4 +3,7 @@ not.null.field=This field cannot be left empty
...
+date.taken=This date and hour has been taken by
+date.and.hour.available=The selected date and hour were available
```
Comment (SendyCruz) [inline consultorio-medico/src/main/resources/ValidationMessages.properties]: es que no estaban disponibles?

**Review by SendyCruz — CHANGES_REQUESTED** [review]

File: consultorio-medico/src/main/resources/ValidationMessages.properties (hunk 25 above)
Comment (DavidVelaz17) [reply consultorio-medico/src/main/resources/ValidationMessages.properties]: No, es solo para decir que si se realizó la acción

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/restcontrollers/MedicineController.java [hunk 26]
```diff
@@ -8,42 +8,55 @@
...
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.ModelAttribute;
+import org.springframework.web.bind.annotation.PathVariable;
 
 import java.util.List;
 
 @RestController
-@RequestMapping(path = "/iwaid/medicines")
+@RequestMapping(path = "/iwaid/medicines/")
```
Comment (cgarcia-iwa) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/restcontrollers/MedicineController.java:24]: determinar con el resto del equipo el estandar que utilizaran para las rutas

File: consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/AppointmentService.java [hunk 27]
```diff
@@ -0,0 +1,143 @@
...
+            log.info(messages.getString("date.and.hour.available"));
+    }
+
+    private void validateDuplicatedAppointment(
+    final Integer doctorId,
+    final Integer patientId,
+    final LocalDate date,
+    final Hour hour)
+         throws Exception{
+        if(appointmentRepository.existsByDoctorIdAndPatientIdAndDateAndHour(doctorId,patientId,date,hour)){
```
Comment (cgarcia-iwa) [inline consultorio-medico/src/main/java/com/iwa/iwaid/consultoriomedico/services/AppointmentService.java]: Que pasa si solo actualizo las notas, me dejara actualizar o fallara?

**Review by cgarcia-iwa — CHANGES_REQUESTED** [review]

**Review by cgarcia-iwa — APPROVED** [review]

**Pull request merged**